### ISSUE TITLE
mockoon-cli live update

### ISSUE THREAD

**Issue by AKABTB** [mockoon/mockoon#665]
When I am updating the data file, I want mockoon-cli to sense the changes in real time, without restarting the service, like client software does.

**Comment by benlei-gfm** [comment]
I'm trying to compare mock-server and mockoon, and it seems like mock-server has this feature built in.

Would be great to have this also in mockoon to make it a stronger choice.

**Comment by 255kb** [comment]
> I've also noticed that the UI doesn't watch the settings file. As such if I were to make modifications to a route directly on the file, the Mockoon UI wouldn't be aware of it. Then if I do make a modification with the Mockoon UI, it overwrites/loses my manual change.
This has been added in v1.19.0. You must enable it in the settings.
![image](https://user-images.githubusercontent.com/7489814/180140760-893b4b90-4e38-4d80-9306-4c9e92d29191.png)

**Comment by vltansky** [comment]
Where can i find this settings? @benlei-gfm @255kb

**Comment by 255kb** [comment]
> Where can i find this settings? @benlei-gfm @255kb

@vltansky This setting is only for the UI (ctrl+, or top menu, Application -> settings). The CLI currently offers no monitoring.

**Comment by avrezvanov** [comment]
I want this opportunity too. Automatic restart of the CLI.

**Comment by wilcorrea** [comment]
Hey guys. Do you have plans to "live update" feature? Can we help on it?

**Comment by 255kb** [comment]
> Hey guys. Do you have plans to "live update" feature? Can we help on it?

Sure, any help is welcomed. Let me know if you want to work on it, we can discuss the implementation as there are multiple use cases to cover: 
- running the CLI with `--daemon-off` (foreground process)
- running the CLI without `--daemon-off` (PM2 managed process). Can we monitor correctly? Not sure. 
- running the CLI within Docker: 
  - with the provided Docker image, data file is passed at runtime.
  - with a custom Docker image, data file copied inside the container (maybe easier).
 
And for all theses cases, a path to a local data file can be passed as a param, or a URL. Do we monitor URLs by using some polling? Or do we exclude them from the feature?

**Comment by VladimirCores** [comment]
It would be good to have this feature as part of `mockook-cli` API.

**Comment by Kartik-Singla** [comment]
Is this feature still under consideration. This feature would be really useful

**Comment by 255kb** [comment]
> Is this feature still under consideration? This feature would be really useful

It's not on the roadmap yet, but I will move it. If someone wants to contribute they are welcome.

**Comment by realjjaveweb** [comment]
@255kb did this move in your roadmap? it it already being worked on? Do you have any tips on where to start if not?

**Comment by 255kb** [comment]
> [@255kb](https://github.com/255kb) did this move in your roadmap? is it already being worked on? Do you have any tips on where to start if not?

It's on the *massive* list of things I want to do. We are also a bit more focused on our cloud this year (for sustainability reasons, you can read more about this [here](https://mockoon.com/blog/introducing-mockoon-cloud-web-app/#sustainability-and-shifting-focus-to-mockoon-cloud)). So, if you want to contribute, please do it! 

For me the scope of this feature is more or less the following: 

What should be included:

- adding a `--watch` flag to enable file watching (+update the docs).
- using chokidar (we are already using chokidar for the desktop app, I guess it makes sense to stick with the same lib) in the CLI to monitor changes to the environment file (data file) and relaunch (`stop()`/`start()`) the instance after updating the environment object inside the `MockoonServer` class instance.

What should not be included (imo):
 
- monitoring when running in a Docker container.
- it's not realistic to monitor when loading the environment from a URL (we shouldn't poll or something).

**Comment by realjjaveweb** [comment]
@255kb thank you! Can you please also point me to some codebase file (I'm assuming within https://github.com/mockoon/mockoon/tree/main/packages/cli) on where such hooks should be implemented and from where called/initiated?

> monitoring when running in a Docker container.
- wdym? Because I think the most people if they're using cli, they will use it dockerized, so IMO it makes the most sense to do the monitoring _**especially**_ when running in a docker container - typical usecase - you have mounted volume with data files and you would like to edit them with the GUI, but see the results immediatelly in your app - which is - most likely - connected to that docker container. However does it really make any sense to differentiate here? Or did you mean more like "detecting docker" (which doesn't make sense IMO).

**Comment by 255kb** [comment]
> [@255kb](https://github.com/255kb) thank you! Can you please also point me to some codebase file (I'm assuming within https://github.com/mockoon/mockoon/tree/main/packages/cli) on where such hooks should be implemented and from where called/initiated?

Yes, [this part of the code](https://github.com/mockoon/mockoon/blob/main/packages/cli/src/commands/start.ts#L176-L245) is the one managing the MockoonServer instance. I would try to add the file monitoring code there first. 

> > monitoring when running in a Docker container.
> 
> * wdym? Because I think the most people if they're using cli, they will use it dockerized, so IMO it makes the most sense to do the monitoring _**especially**_ when running in a docker container - typical usecase - you have mounted volume with data files and you would like to edit them with the GUI, but see the results immediatelly in your app - which is - most likely - connected to that docker container. However does it really make any sense to differentiate here? Or did you mean more like "detecting docker" (which doesn't make sense IMO).

You are definitely right, monitoring files when mounted in a Docker container makes perfect sense. 
I think I was referring to monitoring files when copied inside a Docker container which, in retrospect, doesn't really make sense anyway...

**Comment by 255kb** [comment]
📦Released in [v9.3.0.](https://mockoon.com/releases/9.3.0/) 

---

Mockoon is fully open-source and maintained with love. If it’s helpful and saves you time, consider [supporting the project](https://mockoon.com/sponsor-us/) ❤️
